### PR TITLE
Add local instance metrics page using new Dataverse metrics API

### PR DESCRIPTION
**Pull request by qqmyers**
Adds a page showing the metrics available from [Dataverse PR 7178](https://github.com/IQSS/dataverse/pull/7178) for a single dataverse or sub-Dataverse within it. See https://github.com/IQSS/dataverse/issues/7177#issuecomment-678428554 for a summary. These all include CSV download buttons (#51).

The existing metrics across dataverses (which I understand may be becoming obsolete), are moved to a /global subdirectory but otherwise work as before except for the addition of two graphs showing the distribution of Dataverse versions in use (#50). (Could be dropped if they are fully replaced.)

### CODE REVIEW

**Comment by qqmyers** [comment]
Images of the new local page from QDR's test instance. The tree widget at the top allows you to get the metrics for just that sub-Dataverse (note QDR calls these sub-Collections - that's configurable). 
image
image
image

File: jquery-3.5.1.min.js
Comment (pdurbin) [inline jquery-3.5.1.min.js:1]: Can we delete this file and just get jQuery from a CDN?

**Review by pdurbin — COMMENTED** [review]
The new plots look great!

My first thought with all of these changes is that we need to bump the release number quite a bit due to incompatible changes. Currently, the next milestone is v0.2.8 but what should we change it to? v0.3.0? v1.0.0? Probably the latter. I saw somewhere a suggestion to move all this code to another repo. That's fine too, I suppose.

I was surprised to read "the existing metrics across dataverses (which I understand may be becoming obsolete)." Is this true? They are becoming obsolete?

The installation at https://dataverse.org/metrics is operated by @donsizemore so we'll definitely want to coordinate with him about when and how to roll out the "global" mode in dataverse-metrics 1.0. I assume "global" will continue to aggregate across any Dataverse 4.9+ installation. If this is not the case, please let us know.

With regard to the plots themselves, I like them but last time they when through several rounds of design review. Is this planned?

I'm a little confused about what's going on with CSV vs TSV. dataverse-metrics has always produced TSV files that the front end (d3plus) visualizes. Now the orginal code (now called "global") continues to use TSV but the new code uses CSV? I assume this is because new endpoints on the Dataverse side produce CSV. Is there also just a general preference for CSV? Is TSV weird?

I also left a technical comment or two in the review itself.

Overall, great work!

See also my review on the related pull request on the main Dataverse repo: https://github.com/IQSS/dataverse/pull/7178

File: jquery-3.5.1.min.js
Comment (qqmyers) [reply jquery-3.5.1.min.js:1]: Hmm - the index.html file already does get it remote, so this must just be from earlier testing.

**Comment by qqmyers** [comment]
**Re: version, repo, and obsolete comment:**
Up to you all how to handle this - I'm happy with whatever version. Or it could be moved to a new repo. Nominally there's some overap still between the local and global metrics (if you run the global ones, you can point to them from the local page) but the ended up being further apart than I initially expected (e.g. since the local one calls the new /monthly endpoints rather than using python to sequentially call the toMonth endpoint and assemble a local cached time series). I could see then getting closer again in the future and they still do look very similar in terms of running through the list of endpoints to create plots, so keeping them together might still be worthwhile.

As for obsolete - as I was looking through this, I was told that Harvard was putting together a new way to look at global metrics which I though I understood meant that the global metrics here might not be a priority. I'm not sure anyone ever said 'obsolete' though, so I may have misunderstood.

**Re:  https://dataverse.org/metrics:** I think the updates @donsizemore recently made to handle the QDR metrics (and errors from any site) is all that's needed for everything to be compatible (so the current page will already work if this PR is merged). If dataverse.org/metrics wants to deploy the new code but only show the global page, I think all that would be needed is to tweak path/dir names. And - the "global" page does continue to aggregate across any Dataverse 4.9+ installation, and adds the version plots to show what's actually deployed out there.

**re: design review:** As these were basically updates to the existing global plots, and because this was a QDR versus GDCC PR, there hasn't been any discussion of further design review or GDCC/IQSS work. That said, the PR is essentially a hand-off, so if IQSS/GDCC want to review further or make changes, before or after merging a base PR, that's all fine. (I'll definitely admit the styling for choosing a sub-Dataverse to show metrics for could use some work. I thought it was worth putting a selector in the page (which is what needed the tree endpoint in the Dataverse PR), but one can go to metrics for a given sub-dataverse directly via URL and the API, so it was a bit of a bonus feature w.r.t. what QDR needed).

re: tsv/csv: I don't think there's much reason to prefer one over the other (though I think CSV is better known?), and it would be easy at this point to add TSV in addition or switch (just need to define the default). That said, I didn't change the Dataverse side from TSV to CSV - the original API produced json-only, with single numbers for the time-series, which are then assembled into TSV by the local python scripts in dataverse-metrics. Also - FWIW: I think the d3plus libraries can handle either, so if there ever was a change to use the direct time-series calls in the new API, I think the global code could use the same code (if it didn't make sense to just switch to use the json endpoints as the local page does).

**Comment by qqmyers** [comment]
and thanks again for looking through all this!

**Comment by pdurbin** [comment]
In standup yesterday I said I'd re-review this pull request. I haven't run the code and not much has changed but I'm not quite ready to send it to QA.

@qqmyers can you please resolve the merge conflicts?

@djbrooke and @TaniaSchlatter any plans to look at this from a design perspective? My understanding is that the "single installation" mode should look more or less like https://data.qdr.syr.edu/metrics_5ef2ae2be4b/

@djbrooke there are enough changes here that I'd suggest v1.0.0 as the next release number.

@donsizemore is it safe to assume you'll be updating https://metrics.dataverse.org once this goes through QA? There we'd use the new "global" mode that's being introduced.

**Comment by qqmyers** [comment]
~Merged. The conflicts were with the config.json.sample, download.py, and plots.js files that were moved to ./global and then changed in both places. The list of sites in global/config.json.sample is the same as Don had in the master branch. For download.py, https://github.com/IQSS/dataverse-metrics/commit/01074f21579caf09e47749d372083dc4ea7a4b1f shows the minor difference in error handling. 
Note that the main thing that happened with the old code is that it moved to the ./global subdir. Otherwise, the only real changes is the addition of a plot of Dataverse versions found.

**Review by pdurbin — APPROVED** [review]
I haven't run the code but I think it looks ok.

@djbrooke I'm leaving you assigned for now. The two outstanding questions are:

- Do we want any kind of design review?
- When we release, should we bump the version to v.1.0.0? We can decide this later, of course.

**Comment by djbrooke** [comment]
Thanks @qqmyers and @pdurbin. Sorry if I missed this, but is there an option to keep the view that we currently have on dataverse.org/metrics (screenshot attached) or will we move to this new view automatically? I'd just like to understand what the options are.  

I'll defer to @TaniaSchlatter on the design review. We have https://www.iq.harvard.edu/metrics as well, so there may also be a question of where we want to spend time reviewing/changing views for the similar metrics.

Screen Shot 2021-11-30 at 3 55 00 PM

**Comment by qqmyers** [comment]
The current metrics app is still 'as is' in the new branch, but it has moved to the ./global subdirectory. So, the options to keep things as they are for a given deployment are: just don't update, or update and change the relevant paths/links/cron jobs to include /global (or use symlinks/change docroot etc. - basically adapt to the new subdir). One doesn't have to expose the local app at all.

The new app, taking over the base URL (e.g. what would be https://metrics.dataverse.org/ with the global one now at https://metrics.dataverse.org/global/ ) is different in a few ways:

- it's only capturing metrics from the local app
- it is simpler in that it is only html/javascript - no python that has to be run periodically by cron job
- it allows per-sub-collection metrics with a tree widget at the top to select what you want to see.
- if you have configured to run the global app as well, the local app can display a link to it at the bottom.

FWIW: The decision to move the global app to a subdir is ~arbitrary - the local app could go into a local subdir instead. The reason I did it this way because I assumed more sites would want to run the local app to get sub-collection metrics and only some of those would also set up the global app and it would be easier to have them ignore and/or delete the global subdir than to leave the global code with .py files etc. at the root dir.

**Comment by qqmyers** [comment]
Also - the demo from the community call should be around for this somewhere - that should be up-to-date with the PR.

**Comment by TaniaSchlatter** [comment]
The design seems fine for QDR and others who do not have many collections. [Collapsable side navigation](https://www.w3schools.com/howto/howto_js_dropdown_sidenav.asp) would be a good choice. I'm not sure accordions are useful here. They are typically used for displaying sections of information when one is open and the rest are closed by default, to help a user focus on one thing at a time. We do not follow this best practice on Dataverse, but it's not too late to start ;) Charts could be broken up by text headers if all are displayed by default.

**Pull request merged**
